pkg/api/sidecar/v1.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package v1
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/route"
+	"github.com/thanos-io/thanos/pkg/promclient"
+	"github.com/thanos-io/thanos/pkg/shipper"
+
+	"github.com/thanos-io/thanos/pkg/api"
+	extpromhttp "github.com/thanos-io/thanos/pkg/extprom/http"
+	"github.com/thanos-io/thanos/pkg/logging"
+)
+
+// SidecarAPI is a very simple API used by Thanos Sidecar.
+type SidecarAPI struct {
+	baseAPI *api.BaseAPI
+	client  *promclient.Client
+	shipper *shipper.Shipper
+	promURL *url.URL
+	dataDir string
+	logger  log.Logger
+	reg     prometheus.Registerer
+}
+
+// NewSidecarAPI creates an Thanos Sidecar API.
+func NewSidecarAPI(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	client *promclient.Client,
+	shipper *shipper.Shipper,
+	dataDir string,
+	promURL *url.URL,
+) *SidecarAPI {
+	return &SidecarAPI{
+		baseAPI: api.NewBaseAPI(logger, true, nil),
+		logger:  logger,
+		client:  client,
+		reg:     reg,
+		shipper: shipper,
+		dataDir: dataDir,
+		promURL: promURL,
+	}
+}
+
+func (s *SidecarAPI) Register(r *route.Router, tracer opentracing.Tracer, logger log.Logger, ins extpromhttp.InstrumentationMiddleware, logMiddleware *logging.HTTPServerMiddleware) {
+	s.baseAPI.Register(r, tracer, logger, ins, logMiddleware)
+
+	instr := api.GetInstr(tracer, logger, ins, logMiddleware, true)
+	r.Post("/flush", instr("flush", s.flush))
+}
+
+type flushResponse struct {
+	BlocksUploaded int `json:"blocksUploaded"`
+}
+
+func (s *SidecarAPI) flush(r *http.Request) (interface{}, []error, *api.ApiError, func()) {
+	dir, err := s.client.Snapshot(r.Context(), s.promURL, false)
+
+	if err != nil {
+		return nil, nil, &api.ApiError{Typ: api.ErrorInternal, Err: fmt.Errorf("failed to snapshot: %w", err)}, func() {}
+	}
+
+	snapshotDir := s.dataDir + "/" + dir
+
+	s.shipper.SetDirectoryToSync(snapshotDir)
+	uploaded, err := s.shipper.Sync(r.Context())
+	if err := os.RemoveAll(snapshotDir); err != nil {
+		s.logger.Log("failed to remove snapshot directory", err.Error())
+	}
+	if err != nil {
+		return nil, nil, &api.ApiError{Typ: api.ErrorInternal, Err: fmt.Errorf("failed to upload head block: %w", err)}, func() {}
+	}
+	return &flushResponse{BlocksUploaded: uploaded}, nil, nil, func() {}
```

Comment:
Shouldn't we immediately shutdown Prometheus after this is done? Otherwise, the same risk of overlapping blocks appears.

Reply:
That's a good point - if the endpoint is only meant to be called once before shutting down, does it even need to be an endpoint? Should it just be the default behavior of the sidecar shutting down? That (I think) should eliminate the concern of both overlapping blocks and two shippers racing against one another.

Reply:
I would love it if all of this would be handled automatically by Sidecar when it is shutting down, but that would entail deeper integration between Sidecar & Prometheus. In Sidecar, we should only care about the head block and avoid overlaps. If only there were a way to tell Prometheus to trim data from the head block that had been uploaded by Sidecar already. :(

Maybe an alternative would be to shut down Prometheus and then read the HEAD block ourselves from Sidecar?

Reply:
I quite like that idea - I played around with it locally and it does work! 
https://gist.github.com/Nashluffy/097e7df7d0a90b0cdefd2b87fb3129c8

But there's a few issues with it, mostly that opening a ReadOnlyDB reads in the WAL before it can be persisted to disk, which depending on the size can be very time consuming & memory intense. If we did want to take this approach, I think there'd have to be a finalizer on the prometheus-owned statefulsets, and a controller that would remove the finalizer. The controller would create a pod that mounts the volume, opens a ReadOnlyDB, and then persists to disk & uploads. But this is quite a lot of orchestration.

The more I think about this, the more I think persist-head-to-block needs to be an endpoint on prometheus/tsdb that we consume. There's quite an extensive thread here about it https://github.com/prometheus-junkyard/tsdb/issues/346